sound/soc/intel/boards/sof_sdw_rt_sdca_jack_common.c

```diff
@@ -176,6 +176,7 @@ int sof_sdw_rt_sdca_jack_exit(struct snd_soc_card *card, struct snd_soc_dai_link
 
 	device_remove_software_node(ctx->headset_codec_dev);
 	put_device(ctx->headset_codec_dev);
+	ctx->headset_codec_dev = NULL;
```

Comment:
The reasoning described by the commit message does not sound encouraging,
The sof_sdw_rt_sdca_jack_init() is also called twice? The sdw_dev is the same or different for the calls?
If we have multiple jacks then we should not use a single pointer to store the two headset's information, no?

Reply:
You also need to check above:
```c
if (!ctx || !ctx->headset_codec_dev)
```
I have seen a NULL dereference crash.

Reply:
> The reasoning described by the commit message does not sound encouraging, The sof_sdw_rt_sdca_jack_init() is also called twice? The sdw_dev is the same or different for the calls? If we have multiple jacks then we should not use a single pointer to store the two headset's information, no?

No, `sof_sdw_rt_sdca_jack_init()` will not be called twice. `sof_sdw_rt_sdca_jack_init()` is called when a dai_link is created. [set_codec_init_func](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/intel/boards/sof_sdw.c#L1574). But we call `.exit` in [mc_dailink_exit_loop](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/intel/boards/sof_sdw.c#L1889), and check all .init in codec_info_list[].



> You also need to check above:
> 
> ```c
> if (!ctx || !ctx->headset_codec_dev)
> ```
> 
> I have seen a NULL dereference crash.

No, `ctx->headset_codec_dev` has been checked at the beginning of `sof_sdw_rt_sdca_jack_exit`().

Reply:
I see, the jack detection is only supported for playback?
If we have a line-in jack then we prevent the detection?
```c
diff --git a/sound/soc/intel/boards/sof_sdw_rt_sdca_jack_common.c b/sound/soc/intel/boards/sof_sdw_rt_sdca_jack_common.c
index e430be7681d2..d9c283829fc7 100644
--- a/sound/soc/intel/boards/sof_sdw_rt_sdca_jack_common.c
+++ b/sound/soc/intel/boards/sof_sdw_rt_sdca_jack_common.c
@@ -176,6 +176,7 @@ int sof_sdw_rt_sdca_jack_exit(struct snd_soc_card *card, struct snd_soc_dai_link
 
 	device_remove_software_node(ctx->headset_codec_dev);
 	put_device(ctx->headset_codec_dev);
+	ctx->headset_codec_dev = NULL;
 
 	return 0;
 }
@@ -191,10 +192,10 @@ int sof_sdw_rt_sdca_jack_init(struct snd_soc_card *card,
 	int ret;
 
 	/*
-	 * headset should be initialized once.
-	 * Do it with dai link for playback.
+	 * Jack detection should be only initialized once for headsets since
+	 * the playback/capture is sharing the same jack
 	 */
-	if (!playback)
+	if (ctx->headset_codec_dev)
 		return 0;
 
 	sdw_dev = bus_find_device_by_name(&sdw_bus_type, NULL, dai_links->codecs[0].name);

```